### ISSUE TITLE
Static default value in Union not transformed for container

### ISSUE THREAD

**Issue by kitchoi** [enthought/traits#1131]
If a list is provided as the static default value in `Union`, it does not get transformed into `TraitListObject` and subsequent mutations to the container are not validated. If change notifications are defined, mutations will not fire notifications either.

e.g.
```
from traits.api import *

class Foo(HasTraits):
    values = Union(List(Int), Set(Int), default_value=[1, 2])

foo = Foo()
print(foo.values)   # print [1, 2]
print(type(foo.values))   # print <class 'list'>
foo.values.append("a")     # unexpected: this does NOT raise
```

This issue likely applies to `dict`(transform to `TraitDictObject`) and `set`(transform to `TraitSetObject`) as well.

**Comment by kitchoi** [comment]
Looks like Either suffers from the same problem.  The following test fails:
```

    def test_either_with_static_default_list(self):

        class Foo(HasTraits):

            str_or_list_of_str = Either(Str(), List(Str), default=["now"])

        foo = Foo()
        self.assertEqual(foo.str_or_list_of_str, ["now"])

        with self.assertRaises(TraitError):
            foo.str_or_list_of_str.append(1)
```

Test failure:
```
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traits/traits/tests/test_either.py", line 40, in test_either_with_static_default_list
    foo.str_or_list_of_str.append(1)
AssertionError: TraitError not raised
```

While this would be an orthogonal issue, I suspect there is a common fix, and that common fix may affect more than one trait types.

**Comment by kitchoi** [comment]
I looked into what caused the static default value not to be transformed, and my finding is summarized here: https://github.com/enthought/traits/issues/948#issuecomment-634544133

For Union, it is actually not always possible to perform validation (and transformation) early when the trait type is instantiated, because the validation rule can be dynamic, e.g. consider `Enum` being one of the trait types in `Union`.  Therefore we can't just fix this trait type like we did for `PrefixMap` (#1139).

I believe the right fix should only apply to transformation and should apply to all trait types.  This could be a change with a much wider impact (not sure), but we definite need more discussion on this, and possibly not suitable to be included in 6.1 at the stage.

Unassigning myself from this issue.
